src/Entity/Citation.php

```diff
@@ -0,0 +1,377 @@
+      'genre' => $this->getGenre(),
+      'page' => $this->getPage(),
+      'publisher' => $this->getPublisher(),
+      'publisher-place' => $this->getPublisherPlace(),
+      'subtitle' => $this->getSubtitle(),
+      'type' => $this->getType(),
+      'volume' => (int) $this->getVolume(),
+    ];
+
+    if ($data['type'] == 'article-journal') {
+      $data['collection-title'] = $data['publisher'];
+    }
+
+    // Convert the arrays into objects.
+    $data = json_decode(json_encode([array_filter($data)]));
+
+    $local_csl = __DIR__ . '/Styles/' . $style . '.xml';
+    if (!file_exists($local_csl)) {
+      return '';
+    }
+
+    // Load the style CSL file.
+    $style = file_get_contents($local_csl);
+    $citeProc = new CiteProc($style);
+    return htmlspecialchars_decode($citeProc->render($data));
+  }
+
+  /**
+   * If the citation can be linked to a url, get the first part of the <a> tag.
+   *
+   * @return string|null
+   *   First half of the <a> tag.
+   *
+   * @throws \Drupal\Core\Entity\EntityMalformedException
+   */
+  protected function getLinkBeginning() {
+    if ($link = $this->getLink()) {
+      // This will pull out the `<a href....>` part of the link.
+      preg_match('/<.*?>/', (string) $link->toString(), $matches);
+      return $matches[0] ?? NULL;
+    }
+  }
+
+  /**
+   * Get the label of the entity wrapped in a link tag to the parent or url.
+   *
+   * @return \Drupal\Core\Link|null
+   *   Label or linked label string.
+   *
+   * @throws \Drupal\Core\Entity\EntityMalformedException
+   */
+  protected function getLink() {
+    $url = NULL;
+
+    // Link to the parent node.
+    if ($parent_entity = $this->getParentEntity()) {
+      $url = $parent_entity->toUrl();
+    }
+
+    // The user entered url.
+    if ($url_string = $this->getUrl()) {
+      $url = $this->getUrlFromString($url_string) ?? $url;
+    }
+
+    // The doi field is a string, so construct the full url from that.
+    if ($doi = $this->getDoi()) {
+      $url = $this->getUrlFromString("https://doi.org/$doi") ?? $url;
+    }
+
+    if ($url) {
+      return Link::fromTextAndUrl('[replace]', $url);
+    }
+  }
+
+  /**
+   * Get a url object from the provided url/string.
+   *
+   * @param string $string
+   *   User entered string.
+   *
+   * @return \Drupal\Core\Url|null
+   *   Url object if successful.
+   */
+  protected function getUrlFromString($string): ?Url {
+    try {
+      return Url::fromUserInput($string);
+    }
+    catch (\Exception $e) {
+      try {
+        return Url::fromUri($string);
+      }
+      catch (\Exception $e) {
+        // Nothing to do, just fall back to returning null.
+      }
+    }
+    return NULL;
+  }
+
+  /**
+   * Fallback function to get the entity field's string value.
+   *
+   * @param string $name
+   *   Function name.
+   * @param mixed $args
+   *   Args.
+   *
+   * @return string
+   *   Entity field value as a string.
+   */
+  public function __call($name, $args) {
```

Comment:
Why not use the magic get function?
https://www.php.net/manual/en/language.oop5.overloading.php#object.get

Reply:
suppose that would do basically the same huh?

Reply:
Oh, that might not work. there's already a `get` method. It would probably be tricky/fragile to override  https://git.drupalcode.org/project/drupal/-/blob/9.2.x/core/lib/Drupal/Core/Entity/ContentEntityBase.php#L1047

Reply:
Seems like this function and the ContentEntityBase class __get function have responsibility overlap but take different approaches. I would be nice to consolidate but not worth the effort IMHO.